plugins/messages/app/models/message.rb

```diff
@@ -4,44 +4,46 @@ class Message < ActiveRecord::Base
   belongs_to :sender, :class_name => "User", :foreign_key => "sender_id"
   belongs_to :group, :class_name => "Group", :foreign_key => "group_id"
   belongs_to :reply_to_message, :class_name => "Message", :foreign_key => "reply_to"
+  has_many :message_recipients, dependent: :destroy
+  has_many :recipients, through: :message_recipients, source: :user
 
-  serialize :recipients_ids, Array
   attr_accessor :send_method, :recipient_tokens, :order_id
 
-  scope :pending, -> { where(:email_state => 0) }
-  scope :sent, -> { where(:email_state => 1) }
-  scope :pub, -> { where(:private => false) }
   scope :threads, -> { where(:reply_to => nil) }
   scope :thread, -> (id) { where("id = ? OR reply_to = ?", id, id) }
-
-  # Values for the email_state attribute: :none, :pending, :sent, :failed
-  EMAIL_STATE = {
-    :pending => 0,
-    :sent => 1,
-    :failed => 2
+  scope :readable_for, -> (user) {
+    joins(:message_recipients).where('private = ? OR message_recipients.user_id = ?', false, user.try(&:id))
   }
 
-  validates_presence_of :recipients_ids, :subject, :body
+  validates_presence_of :message_recipients, :subject, :body
   validates_length_of :subject, :in => 1..255
-  validates_inclusion_of :email_state, :in => EMAIL_STATE.values
 
   after_initialize do
+    @additional_recipients ||= []
     @send_method ||= 'recipients'
   end
 
   before_create :create_salt
-  before_validation :clean_up_recipient_ids, :on => :create
+  before_validation :create_message_recipients, on: :create
+
+  def create_message_recipients
+    users = @additional_recipients
+    users += User.undeleted if send_method == 'all'
```

Comment:
This can be many objects, I'd work with `id`s here.

Reply:
are you sure there is now a difference to the previous implementation? There we called `User.undeleted` too...

Reply:
That's right. It should have said `.pluck(:id)`.
I you can make it, please do so, otherwise it's not a step backward.

Reply:
changed it to id usage only